File: benchmarks/v3/scenarios/05-java-saga/with-mcp/application/saga/ReserveInventoryStep.ts
import {
  SagaStep,
  StepResult,
  CompensationResult,
  SagaError,
  SagaContext,
  InventoryService
} from '../../domain';

interface ReserveInventoryResult {
  reservationId: string;
}

/**
 * Saga step for reserving inventory
 */
export class ReserveInventoryStep implements SagaStep<ReserveInventoryResult> {
  readonly name = 'ReserveInventory';

  constructor(private readonly inventoryService: InventoryService) {}

  async execute(
    context: SagaContext
  ): Promise<StepResult<ReserveInventoryResult>> {
    try {
      if (!context.order) {
        return {
          success: false,
          error: new SagaError({
            message: 'Order not found in context',
            code: 'ORDER_NOT_FOUND',
            stepName: this.name,
            isRetryable: false
          })
        };
      }

      const reservationId = await this.inventoryService.reserveInventory(
        context.orderId,
        [...context.order.items]
      );

      return {
        success: true,
        data: { reservationId }
      };
    } catch (error) {
      return {
        success: false,
        error: new SagaError({
          message: `Failed to reserve inventory: ${(error as Error).message}`,
          code: 'INVENTORY_RESERVATION_FAILED',
          stepName: this.name,
          isRetryable: true
        })
      };
    }
  }

  async compensate(context: SagaContext): Promise<CompensationResult> {
    try {
      if (context.inventoryReservationId) {
        await this.inventoryService.releaseInventory(
          context.inventoryReservationId
        );
      }
      return { success: true };
    } catch (error) {
      return {
        success: false,
        error: new SagaError({
          message: `Failed to release inventory: ${(error as Error).message}`,
          code: 'INVENTORY_RELEASE_FAILED',
          stepName: this.name,
          isRetryable: true
        })
      };
    }
  }
}

File: benchmarks/v3/scenarios/05-java-saga/without-mcp/src/main/java/com/example/saga/application/saga/step/ReserveInventoryStep.java
package com.example.saga.application.saga.step;

import com.example.saga.application.port.InventoryService;
import com.example.saga.application.saga.CompensationResult;
import com.example.saga.application.saga.SagaContext;
import com.example.saga.application.saga.SagaStep;
import com.example.saga.application.saga.StepResult;
import com.example.saga.domain.entity.Order;
import com.example.saga.domain.exception.SagaException;

import java.util.ArrayList;

/**
 * Saga step for reserving inventory.
 */
public class ReserveInventoryStep implements SagaStep<String> {

    private static final String STEP_NAME = "ReserveInventory";

    private final InventoryService inventoryService;

    public ReserveInventoryStep(InventoryService inventoryService) {
        this.inventoryService = inventoryService;
    }

    @Override
    public String getName() {
        return STEP_NAME;
    }

    @Override
    public StepResult<String> execute(SagaContext context) {
        try {
            Order order = context.getOrder()
                    .orElseThrow(() -> new SagaException(
                            "Order not found in context",
                            "ORDER_NOT_FOUND",
                            STEP_NAME,
                            false
                    ));

            String reservationId = inventoryService.reserveInventory(
                    context.getOrderId(),
                    new ArrayList<>(order.getItems())
            );

            return StepResult.success(reservationId);
        } catch (SagaException e) {
            return StepResult.failure(e);
        } catch (Exception e) {
            return StepResult.failure(new SagaException(
                    "Failed to reserve inventory: " + e.getMessage(),
                    "INVENTORY_RESERVATION_FAILED",
                    STEP_NAME,
                    true,
                    e
            ));
        }
    }

    @Override
    public CompensationResult compensate(SagaContext context) {
        try {
            context.getInventoryReservationId().ifPresent(
                    inventoryService::releaseInventory
            );
            return CompensationResult.success();
        } catch (Exception e) {
            return CompensationResult.failure(new SagaException(
                    "Failed to release inventory: " + e.getMessage(),
                    "INVENTORY_RELEASE_FAILED",
                    STEP_NAME,
                    true,
                    e
            ));
        }
    }
}
